### DISCUSSION TITLE
Where does the jupyter compilation process look to find the jupyterlab-language-pack-zh-CN while building jupyter?

### DISCUSSION THREAD

**Discussion by JackYinpei** [jupyterlite/jupyterlite#1222] [Q&A] [answered]
Hello, 

I'm trying to deploy jupyter locally and want to include jupyterlab-language-pack-zh-CN as one of the selectable languages. 

However, when I run the compiled results from 'jupyter lite build', there is no Chinese language option available on the language selection menu. 

I encountered a similar issue before with pyodide kernel not working on the webpage, which I fixed by manually copying the labextensions folder into `sys.prefix`. But for this 'jupyterlab-language-pack-zh-CN', I'm not sure where it needs to be manually placed for it to work,

I have tried to find an answer in "https://jupyterlite.readthedocs.io/en/latest/howto/configure/translation.html", but failed.

Thanks in advance

**Comment by jtpio** [comment]
@JackYinpei this sounds like it could be an issue related to how your environment is setup.

Translation bundles should be installed in the same build environment as other `jupyterlite` dependencie, and are found via entry points.

The `jupyterlite/demo` repo defines `jupyterlab-language-pack-zh-CN` as a dependency: https://github.com/jupyterlite/demo/blob/9504f1d59d420e9da84b7fa3a952e19ef84e6677/requirements.txt#L13. Which is then available via the settings menu:

![image](https://github.com/jupyterlite/jupyterlite/assets/591645/0a161e22-1240-4411-a147-9b4cc31aca7e)

So the easier fix would probably be to setup a new virtual environment and make sure the dependency are correctly installed in the environment.

**Reply by JackYinpei** [reply]
> this sounds like it could be an issue related to how your environment is setup.

yes, I think so too

> Translation bundles should be installed in the same build environment as other jupyterlite dependencie, and are found via entry points.

That is what I want to know - where are the entry points for the language pack?

Over the past few days, I have been trying to build a fully functional environment, but kept failing. The only success I had was on my Windows machine, which was able to correctly compile the plugins I wanted. This has been very frustrating for me.😭😭😭😭

I sincerely apologize for the interruption, but I truly cannot find a solution in the articles. If you could kindly provide some methods, I would be extremely grateful.

**Reply by jtpio** [reply, marked as answer]
So the entry points are mostly implementation details. But for reference here is an example of where they are defined in `jupyterlab-language-pack-zh-CN`:

https://github.com/jupyterlab/language-packs/blob/e0e9f898fafd0f76e7238ddae60ed3f58fc1273b/language-packs/jupyterlab-language-pack-zh-CN/pyproject.toml#L23-L24

In JupyterLite they are picked up using `jupyterlab_server`:

https://github.com/jupyterlite/jupyterlite/blob/0a0becafe4b733d0cfca93d22073d2d424ea797d/py/jupyterlite-core/jupyterlite_core/addons/translation.py#L74C52-L74C52

In `jupyterlab_server` this is where language packs are retrieved via `entry_points`: 

https://github.com/jupyterlab/jupyterlab_server/blob/6704accdf0bd664562ed5cb0d5b025735303cf3b/jupyterlab_server/translation_utils.py#L81-L100

Another point to double check would be to make sure you also have the `jupyterlab_server` dependency installed in your build environment so the language packs can be picked up. I realize the [guide in the docs](https://jupyterlite.readthedocs.io/en/latest/howto/configure/translation.html) doesn't mention it, so we could add it for clarity.

**Reply by JackYinpei** [reply]
![image](https://github.com/jupyterlite/jupyterlite/assets/49555788/035b5432-6830-40ac-b514-e0b81d1272cd)
My Windows machine can display Chinese correctly 

I installed jupyterlab-language-pack-zh-CN on Ubuntu, but when I compiled jupyter, there was no Chinese option available for selection in the browser.
![image](https://github.com/jupyterlite/jupyterlite/assets/49555788/43df57c2-e8f2-4ae1-8290-a8d0f65ab5f6) 

The pip installation path on my system is set to `/usr/local/lib/python3.10/dist-packages`. I believe that during the jupyter compilation process, the jupyter build process should look in the `/usr/share/***'`folders to find the language plugin.  I'm not sure if my understanding is incorrect.

**Reply by jtpio** [reply]
> should look in the `/usr/share/***'`folders to find the language plugin

For extensions yes, but for language packs it should normally be via the entry point machinery and not the shared data in `PREFIX/share`.

**Reply by JackYinpei** [reply]
I succeeded. I installed jupyterlab_server and then ran build again, and I was able to use the Chinese language pack.

Thank you very much，So nice of you

Without your help and support, I could not have completed this work. Thank you for your patience and kindness.

I just want to express my gratitude. I am not a native English speaker, so please excuse any language mistakes. I absolutely do not mean anything else. ❤️❤️❤️❤️❤️❤️

**Reply by jtpio** [reply]
Thanks @JackYinpei!

For reference, I opened https://github.com/jupyterlite/jupyterlite/pull/1223 to improve the documentation.
